File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys

sys.setrecursionlimit(100000)

input = sys.stdin.readline

N = int(input())
triangle = []
dp = [[0] * N for _ in range(N)]

for _ in range(N):
    triangle.append(list(map(int, input().split())))

dp[0][0] = triangle[0][0]

# print(dp)
# print(triangle)

for i in range(1, N):
    for j in range(i+1):
        if j == 0:
            dp[i][j] = dp[i - 1][j] + triangle[i][j]
        else:
            dp[i][j] = triangle[i][j] + max(dp[i - 1][j], dp[i - 1][j - 1])

print(max(dp[N-1]))

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    
    for (let i = 0; i < triangle.length; i++) {
        if (i > 0) {
            for (let j = 0; j < i + 1; j++) { 
                if (j === 0) {
                    triangle[i][j] += triangle[i-1][j]
                } else if (j === i) {
                    triangle[i][j] += triangle[i-1][j-1]
                } else {
                    triangle[i][j] += Math.max(triangle[i-1][j-1], triangle[i-1][j])
                }
            }
        }
    }
    
    answer = Math.max(...triangle[triangle.length - 1])
    
    return answer;
}